pkg/controller/pipelinerun/controller_test.go

```diff
@@ -0,0 +1,175 @@
+/*
+Copyright 2018 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package pipelinerun
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/knative/build-pipeline/pkg/apis/pipeline/v1alpha1"
+	clientset "github.com/knative/build-pipeline/pkg/client/clientset/versioned"
+	pi "github.com/knative/build-pipeline/pkg/client/informers/externalversions/pipeline/v1alpha1"
+	listers "github.com/knative/build-pipeline/pkg/client/listers/pipeline/v1alpha1"
+	"github.com/knative/pkg/test/logging"
+	"go.uber.org/zap"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/client-go/kubernetes"
+	fakekubeclientset "k8s.io/client-go/kubernetes/fake"
+	fakerestclientset "k8s.io/client-go/rest/fake"
+	"k8s.io/client-go/tools/record"
+)
+
+func TestProcessWorkItem(t *testing.T) {
+	logging.InitializeLogger(false)
+	client := fakekubeclientset.NewSimpleClientset()
+	pipelineClient := clientset.New(&fakerestclientset.RESTClient{})
+
+	prs := []*v1alpha1.PipelineRun{{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "test-pipeline-run-success",
+			Namespace: "foo",
+		},
+		Spec: v1alpha1.PipelineRunSpec{
+			PipelineRef: v1alpha1.PipelineRef{
+				Name:       "test-pipeline",
+				APIVersion: "a1",
+			},
+		}},
+		{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "test-pipeline-run-fail",
+				Namespace: "foo",
+			},
+			Spec: v1alpha1.PipelineRunSpec{
+				PipelineRef: v1alpha1.PipelineRef{
+					Name:       "test-pipeline-does-not-exist",
+					APIVersion: "a1",
+				},
+			}},
+	}
+
+	ps := []*v1alpha1.Pipeline{{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "test-pipeline",
+			Namespace: "foo",
+		}},
+	}
+
+	tcs := []struct {
+		name           string
+		pipelineRun    string
+		runtimeSuccess bool
+		expectedRecord string
+	}{
+		{"success", "foo/test-pipeline-run-success", true, "Normal Synced Found pipeline test-pipeline"},
+		{"fail", "test-pipeline-run-fail", false, ""},
+	}
+	ch := make(chan string, 10)
+	defer close(ch)
+	setUpMocks(prs, ps, ch)
+
+	c := NewController(client, pipelineClient, logging.GetContextLogger(t.Name()).Logger).(*Controller)
+	for _, tc := range tcs {
+		t.Run(tc.name, func(t *testing.T) {
+			old := runtime.ErrorHandlers
+			var actualErr error
+			defer func() { runtime.ErrorHandlers = old }()
+			runtime.ErrorHandlers = []func(error){
+				func(err error) {
+					actualErr = err
+				},
+			}
```

Comment:
maybe a comment for the block above to explain it's swapping out the errorhandlers and why?